README.rst
Flask-Hookserver
================

.. image:: https://travis-ci.org/nickfrostatx/flask-hookserver.svg?branch=master
    :target: https://travis-ci.org/nickfrostatx/flask-hookserver?branch=master

.. image:: https://coveralls.io/repos/github/nickfrostatx/flask-hookserver/badge.svg?branch=master
    :target: https://coveralls.io/github/nickfrostatx/flask-hookserver?branch=master

.. image:: https://readthedocs.org/projects/flask-hookserver/badge/?version=latest
    :target: https://flask-hookserver.readthedocs.org/en/latest/

.. image:: https://img.shields.io/pypi/v/flask-hookserver.svg
    :target: https://pypi.python.org/pypi/flask-hookserver

.. image:: https://img.shields.io/pypi/l/flask-hookserver.svg
    :target: https://raw.githubusercontent.com/nickfrostatx/flask-hookserver/master/LICENSE

GitHub webhooks using Flask.

This tool receives webhooks from GitHub and passes the data along to a
user-defined function. It validates the HMAC signature, and checks that the
originating IP address comes from the GitHub IP block.

Supports Flask >= 0.9

Installation
------------

.. code-block:: bash

    $ pip install Flask-Hookserver

Usage
-----

.. code-block:: python

    from flask import Flask
    from flask.ext.hookserver import Hooks

    app = Flask(__name__)
    app.config['GITHUB_WEBHOOKS_KEY'] = 'my_secret_key'

    hooks = Hooks(app, url='/hooks')

    @hooks.hook('ping')
    def ping(data, guid):
        return 'pong'

    app.run()

And there we go! ``localhost:8000/hooks`` will now accept GitHub webhook
events.

Config
------

Signature and IP validation are both optional, but turned on by default.  They
can each be turned off with a config flag.

.. code-block:: python

    app = HookServer(__name__)
    app.config['VALIDATE_IP'] = False
    app.config['VALIDATE_SIGNATURE'] = False

If ``VALIDATE_SIGNATURE`` is set to ``True``, you need to supply the secret key
in ``app.config['GITHUB_WEBHOOKS_KEY']``.

Exceptions
----------

If anything goes wrong, a regular ``HTTPException`` will be raised. Possible
errors include:

- 400: Some headers are missing
- 400: The request body isn't valid JSON
- 400: The signature is missing or incorrect
- 403: The request originated from an invalid IP address
- 503: Could not download the valid webhooks IP block from GitHub


HISTORY.rst
.. :changelog:

History
-------

1.1.0 (2016-04-10)
++++++++++++++++++

- Use GITHUB_WEBHOOKS_KEY if provided, fallback on SECRET_KEY

1.0.0 (2015-12-25)
++++++++++++++++++

- Refactor into a proper Flask extension
- Move from hookserver to flask.ext.hookserver
- Introduce a single hooks object that can be registered to multiple apps
- Remove importable blueprint and standalone Flask app
- Rename KEY to GITHUB_WEBHOOKS_KEY

0.3.2 (2015-11-28)
++++++++++++++++++

- Don't actually connect to GitHub from test suite
- Deploy to PyPI from Travis

0.3.1 (2015-11-19)
++++++++++++++++++

- Accept unicode keys
- Default to true for VALIDATE_IP and VALIDATE_SIGNATURE if missing

0.3.0 (2015-11-08)
++++++++++++++++++

- Move hooks URL route into a blueprint
- Report rate limit expire time when the GitHub request limit is exceeded

0.2.1 (2015-10-21)
++++++++++++++++++

- Install Requests security dependencies

0.2.0 (2015-10-21)
++++++++++++++++++

- Proper Python 2.7 support
- Respect GitHub rate limit of 60 requests per hour
- Make key param to HookServer optional
- Add optional param url
- Use VALIDATE_IP and VALIDATE_SIGNATURE instead of DEBUG
- Simplify some error messages
- Add test suite

0.1.4 (2015-04-07)
++++++++++++++++++

- Support IPv4 addresses mapped to IPv6

0.1.3 (2014-07-10)
++++++++++++++++++

- Fix Python 2 bug in IP address decoding

0.1.2 (2014-07-09)
++++++++++++++++++

- Ignore HMAC in debug mode

0.1.1 (2014-07-08)
++++++++++++++++++

- Reload GitHub IP block on each request

0.1.0 (2014-07-07)
++++++++++++++++++

- Initial release


flask_hookserver.py
# -*- coding: utf-8 -*-
"""
Flask-Hookserver: GitHub webhooks using Flask.

:copyright: (c) 2016 by Nick Frost.
:license: MIT, see LICENSE for more details.
"""

from flask import request
from functools import wraps
from werkzeug.exceptions import BadRequest, Forbidden, ServiceUnavailable
import hashlib
import hmac
import ipaddress
import requests
import time
import werkzeug.security

__author__ = 'Nick Frost'
__version__ = '1.0.0'
__license__ = 'MIT'


class Hooks(object):

    """The Hooks object registers handlers to GitHub webhooks events.

    :param app: the optional :class:`~flask.Flask` instance to register
                the extension
    :param url: the url that events will be posted to
    """

    def __init__(self, app=None, url='/hooks'):
        """Initialize the extension."""
        self._hooks = {}
        if app is not None:
            self.init_app(app, url=url)

    def init_app(self, app, url='/hooks'):
        """Register the URL route to the application.

        :param app: the optional :class:`~flask.Flask` instance to
                register the extension
        :param url: the url that events will be posted to
        """
        app.config.setdefault('VALIDATE_IP', True)
        app.config.setdefault('VALIDATE_SIGNATURE', True)

        @app.route(url, methods=['POST'])
        def hook():
            if app.config['VALIDATE_IP']:
                if not is_github_ip(request.remote_addr):
                    raise Forbidden('Requests must originate from GitHub')

            if app.config['VALIDATE_SIGNATURE']:
                key = app.config.get('GITHUB_WEBHOOKS_KEY', app.secret_key)
                signature = request.headers.get('X-Hub-Signature')

                if hasattr(request, 'get_data'):
                    # Werkzeug >= 0.9
                    payload = request.get_data()
                else:
                    payload = request.data

                if not signature:
                    raise BadRequest('Missing signature')

                if not check_signature(signature, key, payload):
                    raise BadRequest('Wrong signature')

            event = request.headers.get('X-GitHub-Event')
            guid = request.headers.get('X-GitHub-Delivery')
            if not event:
                raise BadRequest('Missing header: X-GitHub-Event')
            elif not guid:
                raise BadRequest('Missing header: X-GitHub-Delivery')

            if hasattr(request, 'get_json'):
                # Flask >= 0.10
                data = request.get_json()
            else:
                data = request.json

            if event in self._hooks:
                return self._hooks[event](data, guid)
            else:
                return 'Hook not used\n'

    def register_hook(self, hook_name, fn):
        """Register a function to be called on a GitHub event."""
        if hook_name not in self._hooks:
            self._hooks[hook_name] = fn
        else:
            raise Exception('%s hook already registered' % hook_name)

    def hook(self, hook_name):
        """A decorator that's used to register a new hook handler.

        :param hook_name: the event to handle
        """
        def wrapper(fn):
            self.register_hook(hook_name, fn)
            return fn
        return wrapper


class _timed_memoize(object):

    """Decorator that caches the value of function.

    Does not care about arguments to the function, will still only cache
    one value.
    """

    def __init__(self, timeout):
        """Initialize with timeout in seconds."""
        self.timeout = timeout
        self.last = None
        self.cache = None

    def __call__(self, fn):
        """Create the wrapped function."""
        @wraps(fn)
        def inner(*args, **kwargs):
            if self.last is None or time.time() - self.last > self.timeout:
                self.cache = fn(*args, **kwargs)
                self.last = time.time()
            return self.cache
        return inner


def _load_github_hooks(github_url='https://api.github.com'):
    """Request GitHub's IP block from their API.

    Return the IP network.

    If we detect a rate-limit error, raise an error message stating when
    the rate limit will reset.

    If something else goes wrong, raise a generic 503.
    """
    try:
        resp = requests.get(github_url + '/meta')
        if resp.status_code == 200:
            return resp.json()['hooks']
        else:
            if resp.headers.get('X-RateLimit-Remaining') == '0':
                reset_ts = int(resp.headers['X-RateLimit-Reset'])
                reset_string = time.strftime('%a, %d %b %Y %H:%M:%S GMT',
                                             time.gmtime(reset_ts))
                raise ServiceUnavailable('Rate limited from GitHub until ' +
                                         reset_string)
            else:
                raise ServiceUnavailable('Error reaching GitHub')
    except (KeyError, ValueError, requests.exceptions.ConnectionError):
        raise ServiceUnavailable('Error reaching GitHub')


# So we don't get rate limited
load_github_hooks = _timed_memoize(60)(_load_github_hooks)


def is_github_ip(ip_str):
    """Verify that an IP address is owned by GitHub."""
    if isinstance(ip_str, bytes):
        ip_str = ip_str.decode()

    ip = ipaddress.ip_address(ip_str)
    if ip.version == 6 and ip.ipv4_mapped:
        ip = ip.ipv4_mapped

    for block in load_github_hooks():
        if ip in ipaddress.ip_network(block):
            return True
    return False


def check_signature(signature, key, data):
    """Compute the HMAC signature and test against a given hash."""
    if isinstance(key, type(u'')):
        key = key.encode()

    digest = 'sha1=' + hmac.new(key, data, hashlib.sha1).hexdigest()

    # Covert everything to byte sequences
    if isinstance(digest, type(u'')):
        digest = digest.encode()
    if isinstance(signature, type(u'')):
        signature = signature.encode()

    return werkzeug.security.safe_str_cmp(digest, signature)
